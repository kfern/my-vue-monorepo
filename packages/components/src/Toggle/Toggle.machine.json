{
	"id": "toggleMachine",
	"initial": "inactive",
	"context": {},
	"states": {
		"inactive": {
			"on": {
				"CLICK": {
					"target": "active"
				}
			}
		},
		"active": {
			"on": {
				"CLICK": {
					"target": "inactive"
				}
			}
		}
	}
}
