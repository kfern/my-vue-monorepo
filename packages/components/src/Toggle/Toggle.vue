<!-- Toggle.vue -->
<template>
  <button @click="send('TOGGLE')">
    {{ msg[state.value] }}
  </button>
</template>

<script>
// import { useMachine } from '@xstate/vue'; // With this syntax jest throw an error
const { useMachine } = require("@xstate/vue");
import { createMachine } from "xstate";

// This machine is completely decoupled from Vue
const toggleMachine = createMachine({
  id: "toggle",
  context: {
    /* some data */
  },
  initial: "inactive",
  states: {
    inactive: {
      on: { TOGGLE: "active" },
    },
    active: {
      on: { TOGGLE: "inactive" },
    },
  },
});

export default {
  name: "Toogle",
  props: {
    msg: {
      active: String,
      inactive: String,
    },
  },
  setup() {
    const { state, send } = useMachine(toggleMachine);
    return {
      state,
      send,
    };
  },
};
</script>
