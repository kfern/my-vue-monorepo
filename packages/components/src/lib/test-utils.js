const { Machine } = require("xstate");
const { createModel } = require("@xstate/test");

// @from: https://timdeschryver.dev/blog/generated-tests-with-xstate-and-cypress#configuring-the-states-for-cypress
const addTests = (machineDefinition, tests) => {
  return {
    ...machineDefinition,
    states: Object.entries(machineDefinition.states).reduce(
      (s, [stateKey, stateValue]) => {
        return {
          ...s,
          [stateKey]: {
            ...stateValue,
            meta: {
              ...stateValue.meta,
              test: tests[stateKey],
            },
          },
        };
      },
      {}
    ),
  };
};

export const getTestsModels = (
  machineDefinition,
  checkState,
  triggerEvents
) => {
  const testMachine = Machine(addTests(machineDefinition, checkState));
  const testModel = createModel(testMachine).withEvents(triggerEvents);
  return testModel;
};
