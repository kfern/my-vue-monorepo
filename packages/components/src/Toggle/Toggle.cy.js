import { getTestsModels } from "../lib/test-utils";
import { mount } from "@cypress/vue";
import Toggle from "@/Toggle/Toggle.vue";
import machineJSON from "@/Toggle/Toggle.machine.json";

const now = new Date().toLocaleString();

const testProps = {
  inactive: "Text when inactive " + now,
  active: "Text when active " + now,
};

const testSelectors = {
  states: {
    inactive: {
      class: ".inactive-screen",
      value: testProps.inactive,
    },
    active: {
      class: ".active-screen",
      value: testProps.active,
    },
  },
  events: {
    CLICK: ".click-button",
  },
};

const triggerEvents = {
  CLICK: {
    exec: ({ cy, selectors }) => {
      cy.get(selectors.events.CLICK).click();
    },
  },
};

const checkState = {
  inactive: ({ cy, selectors }) => {
    cy.get(selectors.states.inactive.class).contains(
      selectors.states.inactive.value
    );
  },
  active: ({ cy, selectors }) => {
    cy.get(selectors.states.active.class).contains(
      selectors.states.active.value
    );
  },
};

const testModel = getTestsModels(machineJSON, checkState, triggerEvents);

describe("Toggle", () => {
  const testPlans = testModel.getShortestPathPlans();

  testPlans.forEach((plan) => {
    describe(plan.description, () => {
      plan.paths.forEach((path) => {
        it(path.description, () => {
          // do any setup, then...
          mount(Toggle, {
            props: testProps,
          });
          cy.waitUntil(() => path.test({ cy, selectors: testSelectors }));
        });
      });
    });
  });

  describe("Coverage", () => {
    it("should have full coverage", () => {
      return testModel.testCoverage();
    });
  });
});
