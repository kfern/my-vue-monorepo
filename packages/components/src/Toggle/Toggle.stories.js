import Toggle from './Toggle.vue';

export default {
  title: 'components/Toggle',
  component: Toggle,
  argTypes: {
    msg: {
      active: String,
      inactive: String
    }
  },
};

const Template = (args) => ({
  // Components used in your story `template` are defined in the `components` object
  components: { Toggle },
  // The story's `args` need to be mapped into the template through the `setup()` method
  setup() {
    return { args };
  },
  // And then the `args` are bound to your component with `v-bind="args"`
  template: '<toggle v-bind="args" />',
});

export const Example = Template.bind({});

Example.args = {
  msg : {
    inactive: 'Click to activate',
    active: 'Active! Click to deactivate'
  }
};
