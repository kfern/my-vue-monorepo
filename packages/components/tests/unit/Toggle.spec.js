import { shallowMount, mount } from "@vue/test-utils";
import Toggle from "@/Toggle/Toggle.vue";

const testProps = {
  msg : {
    inactive: 'Button text when inactive',
    active: 'Button text when active'
  }
};

describe("Toggle.vue", () => {
  it("renders inactive msg by default", () => {
    const wrapper = shallowMount(Toggle, {
      props: testProps,
    });
    expect(wrapper.text()).toMatch(testProps.msg.inactive);
  });

  it("renders active msg When inactive and click", async () => {
    const wrapper = shallowMount(Toggle, {
      props: testProps,
    });

    // Act
    await wrapper.find('button').trigger('click');

    // Assert
    expect(wrapper.text()).toMatch(testProps.msg.active);
  });

  it("renders inactive msg When active and click", async () => {
    const wrapper = shallowMount(Toggle, {
      props: testProps,
    });
    await wrapper.find('button').trigger('click'); // Go to active
    expect(wrapper.text()).toMatch(testProps.msg.active); // check

    // Act. Go to active
    await wrapper.find('button').trigger('click');

    // Assert
    expect(wrapper.text()).toMatch(testProps.msg.inactive);
  });  

  it("is stable", () => {
    const wrapper = shallowMount(Toggle, {
      props: testProps,
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
