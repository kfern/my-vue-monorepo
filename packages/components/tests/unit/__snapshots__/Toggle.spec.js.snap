// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle.vue is stable 1`] = `<button>Button text when inactive</button>`;
